File: src/main/java/io/diekema/dingo/feast/destinations/FileSystemDestination.java
package io.diekema.dingo.feast.destinations;

import io.diekema.dingo.feast.Asset;
import io.diekema.dingo.feast.Exchange;
import io.diekema.dingo.feast.Features;
import org.apache.commons.lang3.StringUtils;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

/**
 * Created by rdiekema on 9/23/16.
 */
public class FileSystemDestination implements Destination {

    String outputPath;

    public FileSystemDestination(String outputPath) {
        this.outputPath = outputPath;
    }

    @Override
    public boolean deliver(Exchange exchange) throws IOException {
        if (!StringUtils.isEmpty(outputPath)) {

            for (Asset asset : exchange.getAssets()) {
                Path absoluteOutPut = Paths.get(outputPath).toAbsolutePath();

                if (!Files.exists(absoluteOutPut)) {
                    Files.createDirectories(absoluteOutPut);
                }

                String absolutePath = absoluteOutPut.toString() + File.separator + asset.getName() + Features.DOT + asset.getExtension();
                FileWriter fileWriter = new FileWriter(absolutePath);
                fileWriter.write(asset.getContent());
                fileWriter.flush();
                fileWriter.close();
                asset.setCurrentPath(Paths.get(absolutePath));
            }
        }
        return false;
    }
}

File: src/main/kotlin/io/diekema/dingo/feast/destinations/FileSystemDestination.kt
package io.diekema.dingo.feast.destinations

import io.diekema.dingo.feast.Exchange
import io.diekema.dingo.feast.Features
import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.runBlocking
import org.apache.commons.lang3.StringUtils
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths

/**
 * Created by rdiekema on 9/23/16.
 */
class FileSystemDestination(internal var outputPath: String) : Destination {

    @Throws(IOException::class)
    override fun deliver(exchange: Exchange): Boolean {
        if (!StringUtils.isEmpty(outputPath)) {
            runBlocking {
                exchange.assets.map {
                    async(CommonPool) {
                        val absoluteOutPut = Paths.get(outputPath).toAbsolutePath()

                        if (!Files.exists(absoluteOutPut)) {
                            Files.createDirectories(absoluteOutPut)
                        }

                        val absolutePath = absoluteOutPut.toString() + File.separator + it.name + Features.DOT + it.extension
                        val fileWriter = FileWriter(absolutePath)
                        fileWriter.write(it.content!!)
                        fileWriter.flush()
                        fileWriter.close()
                        it.currentPath = Paths.get(absolutePath)
                    }
                }
            }

            return true
        }
        return false
    }
}
